### ISSUE TITLE
Considering replacing ruby-build with ruby-install and check for binary

### ISSUE THREAD

**Issue by seivan** [asdf-vm/asdf-ruby#204]
> macOS Catalina, asdf 0.8.0 installed via homebrew. 

### Current Behaviour
Installing `ruby 3.0.0` will not only fetch `ruby-build`, but also build `openssl@1.1.`. 
Unless before hand I have this set up. 
```zsh
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)"
```

and make sure this dir contains ruby-build.  
```zsh
ruby_build_dir() {
    echo "$(asdf_ruby_plugin_path)/ruby-build"
}
```

### Expected behavior 
It should lookup to see if `ruby-build` is available in path, maybe it's installed from `Homebrew` before fetching it.
It should also check for dependencies installed via various package managers instead of building from scratch. 

I find the sane defaults used by `ruby-install` to lookup dependencies via the local package manager to be a better option than to require pre-setup before you can use this plugin. 

> ruby-build does not check for system dependencies before downloading and attempting to compile the Ruby source. Please ensure that all requisite libraries are available on your system.

I find this to be poor user experience, especially since it requires up front work and maintaining it yourself with changes. 

PR in https://github.com/asdf-vm/asdf-ruby/pull/205

**Comment by seivan** [comment]
I took a stab at this for myself, if it's interesting, I made a PR. 
Not a fan of having to setup all the necessary linking flags (not just openssl!) each time or having them as an export to build from scratch when they're available already.

**Comment by Stratus3D** [comment]
Thanks for posting about this. ruby-install looks promising, but does it allow us to install all the Ruby versions that ruby-build does?

I found this article https://nts.strzibny.name/ruby-build-and-ruby-install/ by @strzibny that says:

> I would stick to ruby-build if:
>
>    * I am already a user of rbenv or asdf for development
>    * or need to be building bleeding edge rubies

I am not sure if our users are installing "bleeding edge rubies" with asdf-ruby, but if so we may want to stick with ruby-build.

ruby-install does look promising though, and I'm open to whatever is best.

**Comment by seivan** [comment]
@Stratus3D Afaik it still offers preview versions, if those aren't "bleed" enough you can always supply your own mirror. 

I haven't added support for that, because it seems like the wrong option for a tool like `asdf`, as you said, I don't know if the users would do stuff like that. 

But assuming they do, there's always `-M https://bleeding.edge.rubycom`, if it's important, I can add it, I just don't want to clutter the interface.

**Comment by seivan** [comment]
On current master, you can't pick anything that's not offered by `ruby-build`
```shell
  if [ "$install_type" != "version" ]; then
    echoerr "Cannot install specific ref from source, sorry."
    echoerr "For a list of available versions, see \`asdf list-all ruby\`"
    exit 1
  fi
```

I'm running my own fork, could you check bleeding edge versions are offered via `$ asdf list-all ruby`? 

Here's the one from `ruby-install`

```
ruby:
2.5.8
2.6.6
2.7.2
3.0.0
jruby:
9.2.16.0
rbx:
5.0
truffleruby:
21.0.0
truffleruby-graalvm:
21.0.0
mruby:
2.1.2
```

**Comment by strzibny** [comment]
I am the author of the linked post. As a Rubyist I would prefer asdf to use ruby-build as well. I think the tradeoff is better. It's quite rare needing a bleeding edge Ruby beyond released previews, especially for a polyglot user of asdf. +1 from me.

**Comment by seivan** [comment]
> I am the author of the linked post. As a Rubyist I would prefer asdf to use ruby-build as well. I think the tradeoff is better. It's quite rare needing a bleeding edge Ruby beyond released previews, especially for a polyglot user of asdf. +1 from me.

I’m sorry, I don’t follow. I’d assume for polyglots (as myself) you’d be interested in something that’s generic, works for 99% and doesn’t rely on esoteric configurations and environment variables. This change adds support for out of the box support. 

That being said, it adds support for out of the box support only for macOS, I can’t vouch for other platforms but it doesn’t take anything away from those platforms, only adds support for streamlined macOS support.

**Comment by seivan** [comment]
In fact one thing holding asdf back is the massive pre-configuration needed for various languages and platforms, I really dislike they and would like to streamline as much as possible, but I’d just be interested in adding those for macOS. I’ve done the same for Elixir and Erlang, but haven’t pushed yet, might just do that now to show alternative plugin forks at least.

**Comment by seivan** [comment]
@Stratus3D Ping

**Comment by pboling** [comment]
@strzibny Ruby 2.7.2 is not bleeding edge, and I still can't install it with ruby-build unless I use the master branch hack, which took me hours to discover, and almost made me drop asdf entirely.

Based on that experience alone it seems to me that ruby-build isn't being maintained, and an alternative should be considered.

**Comment by strzibny** [comment]
@pboling I meant ruby-install, this change is about ruby-install replacing ruby-build, and I wanted to write I support this change!

My brain wrote ruby-build unfortunately, sorry for the confusion.

**Comment by seivan** [comment]
@pboling @strzibny I've added prebuilt binaries for macOS that will fallback to `ruby-install` if there is no match between `ruby_version` <> `macOS Major` <> `x64 | arm` - Feel free to try it out. 

One of the issues with Ruby is that `LOADPATH` is hardcoded into the binary. So you really need to set up `$RUBYLIB`.
I managed to squeeze it in unless you guys got a better idea. 

```shell
load_os=("$install_path/lib/ruby/$version/"*"-darwin19")
export LOAD_PATH="${load_os[@]}:$install_path/lib/ruby/$version:$LOAD_PATH"
```

**Comment by eregon** [comment]
> @strzibny Ruby 2.7.2 is not bleeding edge, and I still can't install it with ruby-build unless I use the master branch hack, which took me hours to discover, and almost made me drop asdf entirely.
> 
> Based on that experience alone it seems to me that ruby-build isn't being maintained, and an alternative should be considered.

This seems extremely inaccurate. I won't go in details why here, it's clear on https://eregon.me/blog/2021/06/04/review-of-ruby-installers-and-switchers.html

`ruby-install` does not support `-dev` versions at all, i.e., it just does not work and it's part of the intended "anti-features": https://github.com/postmodern/ruby-install/#anti-features. So that would be a big regression.
ruby-build has `ruby-dev`, `jruby-dev`, `truffleruby-dev`, etc. Those are very useful when checking if a bug exists/was fixed on the latest development build of that Ruby.

I understand the choice to install openssl by default on macOS might be not optimal for all cases, please discuss that at https://github.com/rbenv/ruby-build

**Comment by pboling** [comment]
@seivan After further research, I now agree with you completely.  Thanks for pushing back on this idea!  Switching to ruby-install would be a regression. ref: http://disq.us/p/2hmg75o

I do think it would be beneficial to streamline the updating of ruby-build: https://github.com/asdf-vm/asdf-ruby/issues/217

**Comment by Stratus3D** [comment]
@pboling just like with a package management system (Bundle, NPM, etc...), dependencies are best locked to specific version to prevent a dependency from breaking something. I think asdf-ruby is important and I wouldn't want it to break for everyone immediately if ruby-build was changed or someone accidentally tagged a bad commit. I think the slight delay of having to update a line of code to make new rubies available is worth the cost.

**Comment by eregon** [comment]
> I think the slight delay of having to update a line of code to make new rubies available is worth the cost.

The "slight delay" was weeks or months from what I've seen on this issue tracker, e.g., https://github.com/asdf-vm/asdf-ruby/issues/217
I really don't think the tiny risk is worth getting new Ruby releases so much later.
Also, AFAIK `ruby-build` never broke so far, and if it did picking a release wouldn't necessarily help (unless someone would test installing every ruby version with asdf, which is most likely unrealistic).
`ruby-build` is not a library, the command-line interface is fairly limited and I believe a basic usage like `ruby-build version prefix` will always be supported.
It's probably best to discuss this on https://github.com/asdf-vm/asdf-ruby/issues/217 though.

**Comment by alexanderadam** [comment]
> `ruby-install` does not support `-dev` versions at all, i.e., it just does not work and it's part of the intended "anti-features": https://github.com/postmodern/ruby-install/#anti-features. So that would be a big regression.

`ruby-build` basically has the same issue for anyone with a system where installing old dependencies might be difficult but an _old_ Ruby version should be installed, no? :thinking: 

So I guess it would be nice to have the possibility to chose the source where a Ruby version (`ruby-build` or `ruby-install`) should be installed from.

On newer Ubuntu (newer than 18.04) the package `libssl1.0-dev` is not available anymore from sources. [Ubuntu 20.04 only has `libssl-dev` on `1.1.1`](https://packages.ubuntu.com/focal/libssl-dev) and on my tests with Ruby `2.3.8` it didn't build old packages with `asdf-ruby`.

So [this tip in the wiki](https://github.com/asdf-vm/asdf-ruby/wiki/Ruby-Installation-Problems#installing-ruby-23x-on-ubuntu) doesn't seem to work on newer versions.

I understand that installing outdated and unmaintained Ruby as well as installing current bleeding edge versions is rather a corner case (since most people probably need current _stable_ releases anyway) but it would still be nice to have a solution for that if possible.

Currently there doesn't seem to be an easy solution (at least [nobody was able to propose an easy solution on Twitter](https://twitter.com/alexanderadam__/status/1402002096413675529)).

**Comment by eregon** [comment]
> `ruby-build` basically has the same issue for anyone with a system where installing old dependencies might be difficult but an _old_ Ruby version should be installed, no? thinking

That's unrelated. It's possible to install many `-dev` versions with `ruby-build`. With `ruby-install` it's possible to install exactly 0 dev versions of Ruby implementations.

The OpenSSL issue you have is unrelated to ruby-install or ruby-build, in other words using ruby-install won't solve your issue at all.
https://github.com/eregon/ruby-build/tree/ruby23-openssl-linux might help, as it builds an old libssl 1.0 from source. However don't expect any database to work if your system libssl is != Ruby's linked libssl.
Basically, don't use old Rubies requiring libssl 1.0, or if you must use an old OS as well.

**Comment by alexanderadam** [comment]
Ah okay. Thank you for the fast response.

Is there any chance to place precompiled ruby versions so that `asdf-ruby` would recognize it (since placing it to `~/.asdf/installs/ruby/` doesn't seem to be enough)?

**Comment by eregon** [comment]
I missed the initial context of this issue.
Anyway, 2.3.8 on Ubuntu > 18.04 will need to manually compile openssl (or https://github.com/eregon/ruby-build/tree/ruby23-openssl-linux) and it won't work with databases or other system package depending on libssl.
Using ruby-install would use the package manager, and the package manager would tell you there is no libssl1.0 for good reasons: it's unmaintained, unsecure and ancient, i.e., wouldn't solve anything in your case.

Regarding your question I have no idea but it seems better to ask in a separate issue or so.
